cranelift/codegen/src/take_and_replace.rs

```diff
@@ -0,0 +1,91 @@
+//! Helper for temporarily taking values out and then putting them back in.
+
+/// An RAII type to temporarily take a `U` out of a `T` and then put it back
+/// again on drop.
+///
+/// This allows you to split borrows, if necessary, to satisfy the borrow
+/// checker.
+///
+/// The `F` type parameter must project from the container type `T` to its `U`
+/// that we want to temporarily take out of it.
+///
+/// # Example
+///
+/// ```
+/// use cranelift_codegen::TakeAndReplace;
+///
+/// #[derive(Default)]
+/// struct BigContextStruct {
+///     items: Vec<u32>,
+///     count: usize,
+/// }
+///
+/// impl BigContextStruct {
+///     fn handle_item(&mut self, item: u32) {
+///         self.count += 1;
+///         println!("Handled {item}!");
+///     }
+/// }
+///
+/// let mut ctx = BigContextStruct::default();
+/// ctx.items.extend([42, 1337, 1312]);
+///
+/// {
+///     // Temporarily take `self.items` out of `ctx`.
+///     let mut guard = TakeAndReplace::new(&mut ctx, |ctx| &mut ctx.items);
+///     let (ctx, items) = guard.get();
+///
+///     // Now we can both borrow/iterate/mutate `items` and call `&mut self` helper
+///     // methods on `ctx`. This would not otherwise be possible if we didn't split
+///     // the borrows, since Rust's borrow checker doesn't see through methods and
+///     // know that `handle_item` doesn't use `self.items`.
+///     for item in items.drain(..) {
+///         ctx.handle_item(item);
+///     }
+/// }
+///
+/// // When `guard` is dropped, `items` is replaced in `ctx`, allowing us to
+/// // reuse its capacity and avoid future allocations.  ```
+/// assert!(ctx.items.capacity() >= 3);
+/// ```
+pub struct TakeAndReplace<T, U, F>
+where
+    F: Fn(&mut T) -> &mut U,
+    U: Default,
+{
+    container: T,
+    value: U,
+    proj: F,
+}
+
+impl<T, U, F> Drop for TakeAndReplace<T, U, F>
+where
+    F: Fn(&mut T) -> &mut U,
+    U: Default,
+{
+    fn drop(&mut self) {
+        *(self.proj)(&mut self.container) = std::mem::take(&mut self.value);
+    }
+}
+
+impl<T, U, F> TakeAndReplace<T, U, F>
+where
+    F: Fn(&mut T) -> &mut U,
+    U: Default,
+{
+    /// Create a new `TakeAndReplace` that temporarily takes out
+    /// `proj(container)`.
+    pub fn new(mut container: T, proj: F) -> Self {
```

Comment:
Perhaps `container: &mut T` here to emphasize that this isn't intended to take `T` by ownership? (above it's only instantiated as `&mut Self` which means it should still compile I think)